zebra/zebra_ns.c

```diff
@@ -175,19 +179,26 @@ int zebra_ns_final_shutdown(struct ns *ns)
 
 int zebra_ns_init(const char *optional_default_name)
 {
+	struct ns *default_ns;
 	ns_id_t ns_id;
 	ns_id_t ns_id_external;
 
-	dzns = zebra_ns_alloc();
-
 	frr_with_privs(&zserv_privs) {
 		ns_id = zebra_ns_id_get_default();
 	}
 	ns_id_external = ns_map_nsid_with_external(ns_id, true);
 	ns_init_management(ns_id_external, ns_id);
 
+	default_ns = ns_lookup(ns_get_default_id());
+	if (!default_ns) {
+		flog_err(EC_ZEBRA_NS_NO_DEFAULT,
+			 "%s: failed to find default ns", __func__);
+		exit(EXIT_FAILURE); /* This is non-recoverable */
+	}
+
 	/* Do any needed per-NS data structure allocation. */
-	dzns->if_table = route_table_init();
+	zebra_ns_new(default_ns);
+	dzns = default_ns->info;
```

Comment:
zebra_ns_new() is called from inside ns_init_management.
so I think it is useless.

Reply:
it does not seem to be called when i step through with gdb.

I assume you mean the HOOK call? That isn't registered till the end of this function.

Reply:
ok, the hook is done after, not before

		ns_add_hook(NS_NEW_HOOK, zebra_ns_new);
should be called whatever the vrf kind.
and be called before ns_init_management.

so there is a move to do in *zebra_ns_init()* function, if you agree.

Reply:
So... im not sure I agree that zebra should subscribe to netns hooks if its not using netns-based vrfs.

Reply:
Unless there is another use-case internally zebra needs netns's for, why would it need that information when not using netns vrfs?

Reply:
that hook is done for netns creation.
maybe zebra_ns_new() could be modified :
if (!is_backend_netns())
  return;

Reply:
I know what the hook is for.

I am saying zebra does not need to subscribe to that hook unless its using netns-based vrfs.

It doesn't need to know anything about netns creation with normal vrfs.

Reply:
so you propose to directly call zebra_ns_new() instead ?
I am not a big fan of this, since your code will not very well abstracted.
Based on netnamespace abstraction, if we use one or more than one netns, we should rely on the same APIs.

is not it possible to "reuse" the API ?
you want to call zebra_ns_new() outside of the hook registration mechnism ?

Reply:
IMO this 'init' path is special enough that it's fine not to expect the entire netns machinery to be involved - we're just ensuring that _something_ is attached to the default pointer.

Reply:
@pguibert6WIND any more thoughts on this? I will add some additional comments about my thinking here.

I am calling the `zebra_ns_new()` function specifically for the purpose of re-using the code but without subscribing to the HOOKS.

Zebra (by default) should not care about network namespaces imo so it shouldn't subscribe to those hooks.

In lieu of that, I elected to call the code that would be called if it did (`zebra_ns_new()`) rather than subscribing to the hook.

> I am not a big fan of this, since your code will not very well abstracted.
Based on netnamespace abstraction, if we use one or more than one netns, we should rely on the same APIs.

It is more abstracted than the code was before. It does rely on the same API, just doesn't subscribe to the hook.

Reply:
wekeep the code as is.
it is ok forme too.
thanks